### ISSUE TITLE
Events should be colored based on their categories

### ISSUE THREAD

**Issue by AlexandreBonneau** [fullcalendar/fullcalendar#4356]
Coloring events based on their categories would be a must.

Most major 'offline' calendar clients support this (ie. korganizer), and it's far easier to add 50+ categories, on the fly, than to have to setup an additional calendar to share (and don't forget to update all the clients to connect to that new superfluous calendar) every time you want to add a new 'category'/color to your calendar.

For info, the calendar apps that support categories coloring managed the coloring problem by allowing to either:
- use only the categories color (and use only the very first category color if multiple categories are present)
- use only the agenda color (this defeat the purpose of this feature request but oh well)
- use the categories color inside (background color), and the agenda color outside (border color)
- use the agenda color inside (background color), and the categories color outside (border color)

tl;dr
Since you can add categories on events and tasks in an `.ics` file, it would be great to be able to define a color for each category, and have it applied on all the events that have that category (on *all* the calendars).

**Comment by DiemenDesign** [comment]
I found I can target events using their data attribute, then add relevant class to indicate event categories that way. I use it to indicate new bookings, confirmed bookings.

**Comment by AlexandreBonneau** [comment]
Ok, but this rely on hacks where you need to use an external js script to extract the categories for each events.

I was thinking about a simple dropdown where you could select or create new categories in the event detail editor, and the category list in the settings with a color picker for each of those.

**Comment by DiemenDesign** [comment]
Are you aware that the API allows you to add classes or styles when creating the event?

**Comment by AlexandreBonneau** [comment]
In my searches, I've found that you can for each event specify the background/border/text color, but I'm thinking about end users here, where a simple GUI would be involved.
Are you referring to that?

**Comment by DiemenDesign** [comment]
Yep, you should still be able to target the element to change the colour of the event using it's data-attribute.

**Comment by AlexandreBonneau** [comment]
@DiemenDesign I'm not sure how you get the categories in data attributes.

I've created an event in my calendar client ([Korganizer](https://userbase.kde.org/KOrganizer)) and assigned the 'Foobar' category to it.
Now, I can see on Nextcloud that the event synchronized correctly. When I select that event, then export it from Nextcloud, I can see that the category has been saved as well `CATEGORIES:Foobar`.

However, on the html side, I can only see:
```html
<a class="fc-day-grid-event fc-h-event fc-event fc-start fc-end fcCalendar-id-JH1W8EAV5OK4X23ZHZ2AJ" style="background-color:#317CCC;border-color:#317CCC;color:#FAFAFA">
	<div class="fc-content">
		<span class="fc-time">13:15</span> <span class="fc-title">Testing the categories</span>
	</div>
</a>
```
...with no data attribute.

Would it be possible to expose the categories in such data attributes? This would allow for a first quick and dirty solution using an extension like Stylus to correctly color the events.

**Comment by DiemenDesign** [comment]
Look for `data-date`. You can see it highlighted in the right pane.
![Screenshot from 2020-05-20 20-52-26](https://user-images.githubusercontent.com/32765473/82438122-e2189a80-9adb-11ea-96d3-30d86bb745a6.png)

**Comment by AlexandreBonneau** [comment]
The only `data-date` you can see in the html, is the date for each cells in that month view.
Since each event could have different categories assigned to them, the `data-categories` attributes should somehow be added to those events, not to the calendar day.

**Comment by AlexandreBonneau** [comment]
Just to be sure, I checked if fullcalendar correctly store the categories in the .ics files.
To do that:
1. I created an event with a category with Korganizer an computer A
2. Computer A synchronize the events with my nextcloud instance (I checked via fullcalendar that the new event was shown)
3. I modified the event title from within fullcalendar
4. Then from another computer B, another Korganizer instance synchronized the change stored within Nextcloud ; the event had the modified title AND the correct category set by the computer A.

So the good news is fullcalendar does not drop the `CATEGORIES` info, so a quick and easy workaround could be implemented pretty quickly in my humble opinion.
You could imagine a very quick 'fix' for *semi-supporting* category colors ; since the categories respect the ICS RFC and are correctly stored and accessible to fullcalendar, perhaps for each event created, you could do something like:
```js
// Given the following categories are stored in the .ics file:
// CATEGORIES:Foobar,Catégories with accent and spaces,Even - more categories,Quux

// Normalize the category names
const categoriesString = "Foobar,Catégories with accent and spaces,Even - more categories,Quux";
const categoriesArray = categoriesString.split(',');
const catNames = categoriesArray.map(c => camelize(c));

// Set the data attribute accordingly
const event = document.querySelector('.container > a');
event.setAttribute('data-categories', catNames);
```

Then, the user could target the relevant event with the following simple CSS rules:
```scss
.container > a {
	color: #333;
	
	&[data-categories*="foobar"] {
		color      : blue;
		font-weight: bold;
	}
	
	&[data-categories*="iii"] {
		color: violet;
	}
}
```
You can see a working example of that code [here](https://codepen.io/AnotherLinuxUser/pen/vYLmbrx?editors=1111).

Please add the categories as data-attributes, pretty please :)

**Comment by arshaw** [comment]
this can be done with event `extendedProps` along with [event rendering hooks](https://fullcalendar.io/docs/event-render-hooks)

**Comment by AlexandreBonneau** [comment]
I'm glad to hear the current fullcalendar allows for event customization, via DOM events.
However I'm not sure why this close that issue, since the css-savvy end user will not have any way to target the categories which are an integral part of the .ics format.

My point would be that anybody with minimal css knowledge can easily install the Stylus extension and customize the display of events for *any* fullcalendar provider.
However the event rendering hooks implies changing on the server side (not sure how) the fullcalendar code (then change it again after the next update).

The initial issue is to display the categories (usually only one by event) on each event, so that any end-user can use that info if needed. I think having a way to do that internally is great, however does not mean this issue should be closed.

Also, this solution would be the very first step toward a better user experience ;
1/ Display the categories on each event so that tech-savvy user could target those with specific CSS rules
2/ Allow non-tech-savvy users to manage the event/tasks categories, with a color picker.

**Comment by AlexandreBonneau** [comment]
Could this be reopened?
I fail to see how event rendering hooks can be used by end users, unfortunately.
